js/src/vm/PortableBaselineInterpret.cpp

```diff
@@ -805,6 +975,116 @@ ICInterpretOps(BaselineFrame* frame, VMFrameManager& frameMgr, State& state,
     DISPATCH_CACHEOP();
   }
 
+  CACHEOP_CASE(GuardNoDenseElements) {
+    ObjOperandId objId = icregs.cacheIRReader.objOperandId();
+    JSObject* obj = reinterpret_cast<JSObject*>(icregs.icVals[objId.id()]);
+    if (obj->as<NativeObject>().getDenseInitializedLength() != 0) {
+      return ICInterpretOpResult::NextIC;
+    }
+    DISPATCH_CACHEOP();
+  }
+
+  CACHEOP_CASE(GuardStringToIndex) {
+    StringOperandId strId = icregs.cacheIRReader.stringOperandId();
+    Int32OperandId resultId = icregs.cacheIRReader.int32OperandId();
+    BOUNDSCHECK(resultId);
+    JSString* str = reinterpret_cast<JSString*>(icregs.icVals[strId.id()]);
+    int32_t result;
+    if (str->hasIndexValue()) {
+      uint32_t index = str->getIndexValue();
+      MOZ_ASSERT(index <= INT32_MAX);
+      result = index;
+    } else {
+      result = GetIndexFromString(str);
+      if (result < 0) {
+        return ICInterpretOpResult::NextIC;
+      }
+    }
+    icregs.icVals[resultId.id()] = result;
+    DISPATCH_CACHEOP();
+  }
+
+  CACHEOP_CASE(GuardStringToInt32) {
+    StringOperandId strId = icregs.cacheIRReader.stringOperandId();
+    Int32OperandId resultId = icregs.cacheIRReader.int32OperandId();
+    BOUNDSCHECK(resultId);
+    JSString* str = reinterpret_cast<JSString*>(icregs.icVals[strId.id()]);
+    int32_t result;
+    // Use indexed value as fast path if possible.
+    if (str->hasIndexValue()) {
+      uint32_t index = str->getIndexValue();
+      MOZ_ASSERT(index <= INT32_MAX);
+      result = index;
+    } else {
+      PUSH_IC_FRAME();
+      if (!GetInt32FromStringPure(cx, str, &result)) {
```

Comment:
The `Pure` name here is a little confusing to me -- elsewhere when we have a "pure" variant of an operation it doesn't take a `cx`, or if it does it only uses it for things like pre-interned symbols, so we can avoid pushing a frame. I read through some of the called code and it seems to bottom out in `StringToNumberPure` -> `StringToNumber` which can convert a rope string into a linear one, which can OOM; and then the compiled variant of all this handles OOM specially.

All of this to say I think we're doing the right thing here by sticking with the usual idioms, but perhaps a comment could remind us here that "pure" is not strictly "pure", and beware :-)

Reply:
It's not my fault :grin:

This is effectively a disassembly of the macro-assembler invocations in `guardStringToInt32` in `js/src/jit/MacroAssembler.cpp` back to C++. That function generates a call to `GetInt32FromStringPure` without any comments about why that's the right function to call. I'm hesitant to make up a comment without understanding why the existing CacheIR compilers work this way or why they named this function `Pure` while still requiring a context pointer.

I do see that the macro-assembler doesn't call `pushFrameDescriptor` in this case, which I guess means we could get away with calling `cxForLocalUseOnly` instead of pushing an exit frame, right? I hadn't run into `pushFrameDescriptor` before and just found it while writing this response, so I think none of the cases where I added a `PUSH_IC_FRAME` in this PR actually push a frame descriptor in baseline. I'd be happy to double-check that and stop pushing exit frames in all of those cases if you like.

Digging through the call tree of `GetInt32FromStringPure`, it looks to me like we can actually pass a null pointer for `cx` and get exactly the same result. The context is only used to store a flag indicating if we've run out of memory, and the `Pure` path just clears that flag on the way out. But that would mean diverging from how baseline calls this so I don't think we should go that far.

Reply:
Oh, no, the main point of my comment above was the last paragraph -- this is a perfectly fine implementation, and I was just hoping for a one-liner like `// Note that this function, despite the name, takes a `cx` and so we push a frame here as is our idiom.`

That said, you could test using `cxForLocalUseOnly`; I agree with your reasoning that it seems lie it should work, and it's a small but possibly worthwhile optimization if you want to try it.

Reply:
This PR still passes the jit-tests with `cxForLocalUseOnly` in the three cases so I'm going for it!

Reply:
Nice, thanks!